nomad/deployment_endpoint.go

```diff
@@ -13,6 +13,30 @@ import (
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// DeploymentPaginationIterator is a wrapper over a go-memdb iterator that
```

Comment:
API musing, feel free to ignore:

I don't _love_ an API where we have to wrap the go-memdb iterator for every struct we want to paginate. But the only alternative that I can see is a new `PageTokenizer` interface implemented by `Deployment`, `Evaluation`, etc. And then `Paginator` could expect all the `ResultIterators` return results that are `PageTokenizers`. So that's probably not a good path to go down?

Or maybe `Paginator` could take a `ResultIterator` and a function to get the token (it'd close over `byCreateIndex`)? That'd be called something like:

```go
paginator, err := state.NewPaginator(iter, args.QueryOptions,
				func(raw interface{}) error {
					deploy := raw.(*structs.Deployment)
					deploys = append(deploys, deploy)
					return nil
				},
				func(raw interface{}) string {
					deploy := raw.(*structs.Deployment)
					token := deploy.ID
					if byCreateIndex {
							token = fmt.Sprintf("%v-%v", deploy.CreateIndex, deploy.ID)
					}
				},
)
```

But maybe having two function parameters starts calling into question whether `Paginator` should be generic over the type intended to be returned by the `ResultIterator`. (Oh no, the G-word! 😀 )

Reply:
Yeah actually I think generics could clean this up quite a bit. Maybe 1.4 hackathon 🥳

Reply:
>  a new PageTokenizer interface implemented by Deployment, Evaluation, etc. And then Paginator could expect all the ResultIterators return results that are PageTokenizers.

This is similar to the approach we had before, with the `GetID` right? The problem I found is that the struct doesn't know about the request details, so it can't tell if it should `ID` or `CreateIndex` unless we add a new field, which sounded not great.

> Or maybe Paginator could take a ResultIterator and a function to get the token

This was also something that I experimented with, but it felt like I was writing JavasScript 😅 

Another interface I tried was to have a separate method in the `Iterator`, so like:

```go
type Iterator interface {
	Next() interface{}
        GetToken(interface{}) string
}
```

But that didn't help much because you would still need to wrap the `ResultIterator`.

Maybe we can do it with two interfaces? If the iterator implements an interface with `GetToken` we would use that, otherwise we use `GetID` from the raw element.

Reply:
> The problem I found is that the struct doesn't know about the request details, so it can't tell if it should ID or CreateIndex unless we add a new field, which sounded not great.

That's a good point. We'd have to thread the `CreateIndex` through somehow, which ends up being the same as the "pass a function to get the token" implementation.

> This was also something that I experimented with, but it felt like I was writing JavasScript 😅

😆 

> If the iterator implements an interface

I don't know if you can actually check that without reflection or calling a possibly non-existent method?

In any case, I think what we've got here will get the job done. We can always come back and refine it as we have more paginated API examples. (Ex. if we end up wanting to paginate an API by neither ID nor CreateIndex)

Reply:
> In any case, I think what we've got here will get the job done. We can always come back and refine it as we have more paginated API examples. (Ex. if we end up wanting to paginate an API by neither ID nor CreateIndex)

+1. The two interfaces worked, but it was weird to have two very different ways to integrate with the `Paginator`.